### ISSUE TITLE
In 4.03c, the qemu_mode performance loss is more than 10x

### ISSUE THREAD

**Issue by ele7enxxh** [AFLplusplus/AFLplusplus#1540]
Hi all,

I recently used qemu_mode for binary fuzzing(follow this article: https://blog.attify.com/fuzzing-iot-devices-part-1/ ), however in version 4.03c, the performance loss is more than 10x, and on the other hand, it is about 2 times in the old 4.02c version.

After comparing the code of the two versions of qemu_afl, I suspect that the problem lies in the afl_flush_tb function, when I comment out this function in the new version, everything is good.

This is a very serious bug, you should review the code.

Thans,
Zinuo Han.

**Comment by vanhauser-thc** [comment]
@WorksButNotTested one of your two patches did this, can you please find out which one was it and fix it? thank you!

**Comment by WorksButNotTested** [comment]
The change that was made was to flush the translation cache when the forkserver starts. If we don’t then blocks which have been translated before the qemu configuration (environment variables) has been parsed are reused during execution.

If the fork server is not used, then this will result in blocks being translated twice when they may have been only translated once previously. But if you aren’t using forkserver then chances are performance won’t be great anyway.

If your fuzzer executes code which should be instrumented before the config is parsed, then previously, code may have executed and incorrectly not been instrumented and cached before reaching the forkserver (e.g. libc code perhaps). This will result in greater performance, when it is reused, but will not be what was configured.

Since translated blocks should be communicated by the child back to the parent, then the initial loss of cached blocks should not cause any lasting damage since the parent should keep an updated cache at all times to be inherited by the next fork.

**Comment by WorksButNotTested** [comment]
Are the binaries being fuzzed statically linked?

**Comment by vanhauser-thc** [comment]
@WorksButNotTested If the performance degrades we would have to roll back. Is the forkserver pr the culprit or the other patch? Have you checked?

**Comment by WorksButNotTested** [comment]
I suspect the issue is that the binary is statically linked and code executed before the instrumentation is enabled is being cached without any instrumentation (likely something like memcpy). However, without any excluded ranges it should be instrumented. Therefore we see the performance degrade as a result of things being instrumented per the configuration. Whereas before arbitrary code used both during initialisation and the code under test is wrongly uninstrumented.

**Comment by WorksButNotTested** [comment]
> @WorksButNotTested If the performance degrades we would have to roll back. Is the forkserver pr the culprit or the other patch? Have you checked?

The comments above point clearly to the flushing of the translation cache, rather than the changes to deal with Asan.

**Comment by vanhauser-thc** [comment]
@WorksButNotTested 
```
$ clang -O0 -o t3 test-instr.c
$ AFL_PATH=`pwd` ./afl-fuzz -i in -o out -Q -- ./t3
```
if I use HEAD of dev vs 4.02c I get a 10x speed degredation. please fix asap!

**Comment by WorksButNotTested** [comment]
Hmm. That’s is unexpected for a dynamically linked binary. Just trying to understand possible failure scenarios and none are making sense right now. Am I right in thinking of AFL_ENTRYPOINT is not set, it should still us the fork server and run it when it reaches main?

**Comment by WorksButNotTested** [comment]
I can confirm the issue is caused by this commit.
https://github.com/AFLplusplus/qemuafl/commit/cadd60278b5bd0173e1412ded94c42b1973f8509

If you recall, we discussed this on the discord...
https://discord.com/channels/736989425770168422/737363638851207219/1007282059347230750

We raised this issue...
https://github.com/AFLplusplus/qemuafl/issues/35

And this fix...
https://github.com/AFLplusplus/qemuafl/pull/36

The flush happens only once (so it can't be the flush operation itself being slow) and in theory the cache should get repopulated by this.
https://abiondo.me/2018/09/21/improving-afl-qemu-mode/

It can't be an issue with code being uninstrumented when it shouldn't since only the `.text` section should be instrumented and `main` does not call any other functions in the same executable which have already been called.

However, running with the commit patched out with the flag `AFL_QEMU_DISABLE_CACHE=1` gives very similar performance to when the patch is included. So it appears as though our clearing the cache is somehow preventing it from re-filling?!?

I noted that the flushing of the cache takes this branch which appears to defer it...
https://github.com/AFLplusplus/qemuafl/blob/dc19175a0bf4cf34e19944d084d92f33f26df93d/accel/tcg/translate-all.c#L1534

And that by adding a printf here, `do_tb_flush` is being called repeatedly...
https://github.com/AFLplusplus/qemuafl/blob/dc19175a0bf4cf34e19944d084d92f33f26df93d/accel/tcg/translate-all.c#L1487

I can't bracket our call to `afl_flush_tb` with `start_exclusive` and `end_exclusive` to force an asynchronous flush though since we are inside `cpu_exec` and cause a deadlock.

Perhaps @andreafioraldi  has a suggestion?

**Comment by WorksButNotTested** [comment]
Only thing I can think of is to expose and call `do_tb_flush` directly. The purpose of `start_exclusive` and `end_exclusive` seems to be to provide resource isolation from other virtual CPUs...
https://github.com/AFLplusplus/qemuafl/blob/dc19175a0bf4cf34e19944d084d92f33f26df93d/include/hw/core/cpu.h#L899

Given we are in `afl_forkserver` and the fork-server is expected to be started before any threads have been created (as they will be lost with the fork), we could simply forego the synchornization? Or otherwise confirm there is only a single VCPU before we flush the cache. If there is more than one we could skip the flush and just cry into the logs.

**Comment by andreafioraldi** [comment]
Yes you can assume a single vcpu at the forkserver start.

**Comment by andreafioraldi** [comment]
I cannot debug it atm, I have a couple of busy days, what we can do is to revert the change as it was before if we don't find an immediate solution and then I'll look at it later

**Comment by WorksButNotTested** [comment]
How does this look?
https://github.com/AFLplusplus/qemuafl/pull/39/files

**Comment by vanhauser-thc** [comment]
this fixes the speed bug for me.
@WorksButNotTested thank you! does your test case you did the original PR for still work with this?
@ele7enxxh can you please check this works for you as well?
thank you!

Edit: the changes are in the AFL++ dev branch

**Comment by WorksButNotTested** [comment]
> this fixes the speed bug for me. @WorksButNotTested thank you! does your test case you did the original PR for still work with this? @ele7enxxh can you please check this works for you as well? thank you!
> 
> Edit: the changes are in the AFL++ dev branch

I’m fairly confident it will, because I added printfs to make sure do_tb_flush does not bail early. So the cache should be flushed. I’ll try and put a test together as soon as I get a chance. 

Just wanted to get a fix out for the regression asap as that is the more significant issue (the one I fixed went unnoticed for a while, whereas this one didn’t!).

**Comment by andreafioraldi** [comment]
> How does this look? https://github.com/AFLplusplus/qemuafl/pull/39/files

yes lgtm

**Comment by WorksButNotTested** [comment]
The following demonstrates the issue on `4.02c` and remedy in `dev`. Note that assembly is only used to help us easily determine the excluded range and ensure that our code will repeatedly execute a basic block. Obviously this is a contrived example, but my testing showed approx a 200% difference in performance between the two versions, which should be sufficient to confidently show the issue is fixed.
```c
/*
   american fuzzy lop++ - a trivial program to test the build
   --------------------------------------------------------
   Originally written by Michal Zalewski
   Copyright 2014 Google Inc. All rights reserved.
   Copyright 2019-2022 AFLplusplus Project. All rights reserved.
   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at:
     https://www.apache.org/licenses/LICENSE-2.0
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#ifdef TEST_SHARED_OBJECT
  #define main main_exported
#endif

#define BASE_ADDR 0x4000000000

extern char __executable_start;
extern char my_copy_start;
extern char my_copy_end;

void my_copy(char * dest, char* src, size_t cnt);

static void* get_addr(void * ptr) {
  char * ptr_addr = (char *)ptr;
  char * rva = ((char *)BASE_ADDR) + (ptr_addr - &__executable_start);
  return rva;
}

/* args rdi/rsi/rdx */
asm(
  "my_copy_start:\n"
  ".globl my_copy_start\n"
  "my_copy:\n"
  ".globl my_copy\n"
  "jmp 2f\n"
  "1:\n"
  "movb (%rsi), %al\n"
  "movb %al, (%rdi)\n"
  "dec %rdx\n"
  "inc %rsi\n"
  "inc %rdi\n"
  "2:"
  "test %rdx, %rdx\n"
  "jnz 1b\n"
  "ret\n"
  "my_copy_end:\n"
  ".globl my_copy_end\n"
  );

static void delay() {
  static char input[1UL<<20] = {0};
  static char output[1UL<<20] = {0};
  
  my_copy(output, input, 1UL << 20);
}

int test(void) {
  
  delay();
  
  char  buff[8];

  // we support command line parameter and stdin
  if (read(STDIN_FILENO, buff, sizeof(buff)) < 1) {

      printf("Hum?\n");
      return 1;

  }

  // we support three input cases (plus a 4th if stdin is used but there is no
  // input)
  switch (buff[0]) {

    case '0':
      printf("Looks like a zero to me!\n");
      break;

    case '1':
      printf("Pretty sure that is a one!\n");
      break;

    default:
      printf("Neither one or zero? How quaint!\n");
      break;

  }
  return 0;

}

static void print_buf(char * name, char* ptr, size_t len) {
  printf("%s - ", name);
  for (size_t i = 0; i < len; i++)
  {
    printf("%c ", ptr[i]);  
  }
  printf("\n");
}

static void my_copy_test(void) {
  char src[4] = {'a', 'b', 'c', 'd'};
  char dst[4] = {'w', 'x', 'y', 'z'};
  
  print_buf("src", src, sizeof(src));
  print_buf("dst", dst, sizeof(dst));
 
  my_copy(dst, src, sizeof(src));
  
  print_buf("src", src, sizeof(src));
  print_buf("dst", dst, sizeof(dst));
}

int main(int argc, char **argv) {
  if (argc > 1) {
    printf("test: %p\n", get_addr(test));
    printf("my_copy: %p\n", get_addr(&my_copy));
    printf("my_copy_end: %p\n", get_addr(&my_copy_end));
    return 0;
  } else {
    my_copy_test();
    return test();
  }

}
```
Run with arbitrary parameter to print addresses used to configure AFL++.
```
$ gcc -o t3 test-instr.c
$ ./t3 123
test: 0x400000123d
my_copy: 0x4000001205
my_copy_end: 0x400000121a
```

Use deferred forkserver and exclude our assembly routine from coverage.
```
$ AFL_ENTRYPOINT=0x400000123d \
  AFL_QEMU_EXCLUDE_RANGES=0x4000001205-0x400000121a \
  AFL_PATH=`pwd` \
  ./afl-fuzz \
    -i in \
    -o out \
    -Q \
    -- \
      ./t3

```

**Comment by andreafioraldi** [comment]
great work!

**Comment by WorksButNotTested** [comment]
No worries. Sorry again for the regression.
